rfcs/20190729-keras-preprocessing.md

```diff
@@ -0,0 +1,732 @@
+        follow_links: Whether to follow links inside
+            subdirectories (default: False).
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+
+def dataset_from_directory(
+    self,
+    directory,
+    targets='inferred',
+    target_mode='categorical',
+    class_names='inferred',
+    color_mode='rgb',
+    batch_size=32,
+    shuffle=True,
+    seed=None,
+    follow_links=False,
+    validation_split=None,
+    subset='training'):
+    """Takes a directory and returns a Dataset yielding batches of images.
+
+    # Arguments:
+        directory: Directory where the data is located.
+            If `targets` is "inferred", it should contain
+            subdirectories, each containing images for a class.
+            Otherwise, the directory structure is ignored.
+        targets: Either
+            "inferred" (targets are generated from the directory structure),
+            None (no targets),
+            or a list of integer labels of the same size as the number of image
+            files found in the directory.
+        target_mode:
+            - 'categorical' means that the inferred labels are
+                encoded as a categorical vector (e.g. for categorical_crossentropy).
+            - 'binary' means that the inferred labels (there can be only 2)
+                are encoded as binary scalars (e.g. for binary_crossentropy).
+        class_names: Only valid if "targets" is "inferred". This is the explict
+            list of class names (must match names of subdirectories). Used
+            to control the order of the classes (otherwise alphanumerical order is used).
+        color_mode: One of "grayscale", "rgb", "rgba". Default: "rgb".
+            Whether the images will be converted to
+            have 1, 3, or 4 channels.
+        batch_size: Size of the batches of data (default: 32).
+        shuffle: Whether to shuffle the data (default: True)
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        follow_links: Whether to follow links inside
+            subdirectories (default: False).
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+```
+
+Notes:
+- Resizing (`target_size`) is moved to constructor (`resize`).
+- `validation_split` is moved from constructor to here.
+
+#### Question: how to support image segmentation in a simple way?
+
+**Requirements:**
+- Image loading and image augmentation should be synced across inputs and targets
+- It should be possible to use different standardization preprocessing (outside of augmentation) across inputs and targets
+
+**Proposal:** (TBD)
```

Comment:
My only concern is that this proposal is not general enough to accommodate users. If a user wants to do object detection, for example, he will not use keras-preprocessing.

I can work on a new RFC, but I feel that generator/dataset_from_dataframe could be general enough to support it.

Reply:
Have you checked out the proposal for segmentation? Could we use something similar for object detection? What's missing?

Reply:
Right now the segmentation example works because these are images. In a case where we have boxes or keypoints, we don't have that. Furthermore, to do DA on those, we need to normalize those points between 0,1 or provide the shape of the original image.

Reply:
What would be your preferred API for object detection?

Reply:
We would need something similar to flow_from_dataframe and a way to request some info from the image like the shape.
We would also need the ability to change the way we augment 'samples' (image, boxes, keypoint).
 Then we can support any task with little effort I think?

As I said, we should work on another RFCs as this feature is not in scope with this one.

Maybe I should bring it up during the meeting on Monday?

Reply:
Let's discuss this today.